src/coreclr/jit/emitxarch.cpp

```diff
@@ -478,70 +478,151 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
+            case IF_RRW:
 
-            // Bail if not writing to the right register
-            if (id->idReg1() != reg)
-            {
-                return false;
-            }
+            case IF_RRD_CNS:
+            case IF_RWR_CNS:
+            case IF_RRW_CNS:
+            case IF_RRW_SHF:
+
+            case IF_RRD_RRD:
+            case IF_RWR_RRD:
+            case IF_RRW_RRD:
+            case IF_RRW_RRW:
+            case IF_RRW_RRW_CNS:
+
+            case IF_RWR_RRD_RRD:
+            case IF_RWR_RRD_RRD_CNS:
+
+            case IF_RWR_RRD_RRD_RRD:
+
+            case IF_RRD_MRD:
+            case IF_RWR_MRD:
+            case IF_RRW_MRD:
+            case IF_RRW_MRD_CNS:
+
+            case IF_RWR_RRD_MRD:
+            case IF_RWR_MRD_CNS:
+            case IF_RWR_RRD_MRD_CNS:
+            case IF_RWR_RRD_MRD_RRD:
+            case IF_RWR_MRD_OFF:
 
-            // Bail if movsx, we always have movsx sign extend to 8 bytes
-            if (id->idIns() == INS_movsx)
+            case IF_RRD_SRD:
+            case IF_RWR_SRD:
+            case IF_RRW_SRD:
+            case IF_RRW_SRD_CNS:
+
+            case IF_RWR_RRD_SRD:
+            case IF_RWR_SRD_CNS:
+            case IF_RWR_RRD_SRD_CNS:
+            case IF_RWR_RRD_SRD_RRD:
+
+            case IF_RRD_ARD:
+            case IF_RWR_ARD:
+            case IF_RRW_ARD:
+            case IF_RRW_ARD_CNS:
+
+            case IF_RWR_RRD_ARD:
+            case IF_RWR_ARD_CNS:
+            case IF_RWR_ARD_RRD:
+            case IF_RWR_RRD_ARD_CNS:
+            case IF_RWR_RRD_ARD_RRD:
             {
-                return false;
-            }
+                if (id->idReg1() != reg)
+                {
+                    switch (id->idInsFmt())
+                    {
+                        // Handles instructions who write to two registers.
+                        case IF_RRW_RRW:
+                        case IF_RRW_RRW_CNS:
+                        {
+                            if (id->idReg2() == reg)
+                            {
+                                result = (id->idOpSize() == EA_4BYTE);
+                                return PEEPHOLE_ABORT;
+                            }
+                            break;
+                        }
+
+                        default:
+                            break;
+                    }
+
+                    if (instrHasImplicitRegPairDest(id->idIns()))
+                    {
+                        if (id->idReg2() == reg)
+                        {
+                            result = (id->idOpSize() == EA_4BYTE);
+                            return PEEPHOLE_ABORT;
+                        }
+                    }
+
+                    return PEEPHOLE_CONTINUE;
+                }
+
+                // movsx always sign extends to 8 bytes.
+                if (id->idIns() == INS_movsx)
+                {
+                    return PEEPHOLE_ABORT;
+                }
 
 #ifdef TARGET_AMD64
```

Comment:
Shouldn't this entire function be under `ifdef TARGET_AMD64`? After all, x86 can't have upper 32 bits zero ever?

Reply:
At the moment we don't need to, we only use `AreUpper32BitsZero` once:
```cpp
#ifdef TARGET_64BIT
        case GenIntCastDesc::ZERO_EXTEND_INT:
        case GenIntCastDesc::LOAD_ZERO_EXTEND_INT:
            ins     = INS_mov;
            insSize = 4;
            canSkip = compiler->opts.OptimizationEnabled() && emit->AreUpper32BitsZero(srcReg);
            break;
        case GenIntCastDesc::SIGN_EXTEND_INT:
        case GenIntCastDesc::LOAD_SIGN_EXTEND_INT:
            ins     = INS_movsxd;
            insSize = 4;
            break;
#endif
```

Reply:
We don't need to, but wouldn't it document it better as never applying to x86? There's never a case it will apply to x86 in the future, either, right?